src/mlpack/methods/reinforcement_learning/environment/maze.hpp

```diff
@@ -0,0 +1,406 @@
+
+    //! Dimension of the encoded state.
+    static constexpr size_t dimension = 2;
+
+   private:
+    //! Locally-stored {(row, column)}.
+    arma::vec data;
+  };
+
+  /**
+   * Implementation of action of Maze.
+   */
+  class Action
+  {
+   public:
+   enum actions
+    {
+      left,
+      right,
+      down,
+      up,
+    };
+    
+    // To store the action ( direction to move )
+    Action::actions action;
+
+    // Track the size of the action space.
+    static const size_t size = 4;
+  };
+
+  /**
+   * Construct a maze instance using the given constants.
+   * @param rows number of rows in maze.
+   * @param columns number of columns in maze.
+   * @param maxSteps The number of steps after which the episode
+   *    terminates. If the value is 0, there is no limit.
+   */
+  Maze(const size_t rows = 4,
+        const size_t columns = 4,
+        const size_t maxSteps = 200) :
+      maxSteps(maxSteps),
+      stepsPerformed(0)
+  { 
+    MazeGeneration(rows, columns);
+    
+    directions = std::unordered_map<Action::actions, arma::vec>({
+            { Action::actions::left,  arma::vec({0, -1}) },
+            { Action::actions::right,  arma::vec({0, 1}) },
+            { Action::actions::up,  arma::vec({-1, 0}) },
+            { Action::actions::down,  arma::vec({1, 0}) }
+        });
+
+  }
+
+  /**
+   * Generate Maze using Random Iterative DFS Algorithm with 
+   * 0.4 * rows * column steps
+   * @param rows number of rows in maze.
+   * @param columns number of columns in maze.
+   * 
+   */
+  void MazeGeneration(const size_t rows, const size_t columns)
+  {
+    maze = arma::mat(rows,columns);
+    arma::umat visited;
+    visited.zeros(rows, columns);
+
+    std::vector<arma::ivec> moves = {arma::ivec({0, -1}), arma::ivec({0, 1}), arma::ivec({-1, 0}), arma::ivec({1, 0}) };
+
+    const size_t maxNumberOfMazeSteps = 2 * std::sqrt(rows * columns);
+
+    std::stack<arma::ivec> cellsInPath;
+
+    arma::ivec startingPoint = {arma::randi(arma::distr_param(0, rows-1)), arma::randi(arma::distr_param(0, columns-1))};
+
+    cellsInPath.push(startingPoint);
+
+    size_t numberOfMazeSteps = 0;
+ 
+    // loop till stack is empty
+    while (!cellsInPath.empty())
+    {
+        // Pop a vertex from the stack
+        arma::ivec lastCell = cellsInPath.top();
+        cellsInPath.pop();
+
+        ++numberOfMazeSteps;
+
+        // Break for Max number of maze steps.
+        if(numberOfMazeSteps > maxNumberOfMazeSteps)
+          break;
+ 
+        // Mark the particular row and column as visited.
+        visited(lastCell(0),lastCell(1)) = 1;
+
+        // If visitable nodes are present from that node.
+        bool visitableNodes = false;
+
+        // Randomly choose a direction and check if the adjacent node in 
+        // that direction is visited or not.
+        for (size_t counter = 0; counter < 10 ; ++counter)
+        {    
+            size_t index = arma::randi(arma::distr_param(0, 3));
```

Comment:
Init variable using `const` if it isn't changing.

Reply:
Can you please explain in more detail ?

Reply:
`const int index = arma...`